numba/core/interpreter.py

```diff
@@ -955,6 +956,277 @@ def peep_hole_delete_with_exit(func_ir):
     return func_ir
 
 
+def peep_hole_fuse_dict_add_updates(func_ir):
+    """
+    This rewrite removes d1.update(d2) calls that
+    are between two dictionaries, d1 and d2 in the
+    same basic block. This pattern can appear as a
```

Comment:
I generated this example:

```python3
from numba import njit

@njit
def foo(a):
    d1 = {'a':1,
          'b':2,
          'c':3,
          'd':4,
          'e':5,
          'f':6,
          'g':7,
          'h':8,
          'i':9,
          'p':'change type'}
    if a:
        val = 10
    else:
        val = 100
    d1['j'] = val
    return d1['a']

foo(False)
```

Based on this there are a couple of things I noticed (before the peephole optimzations run).
1. The key is defined after the branch (as opposed to before). This may not be consequential.
2. The assignment comes from an explicit `ir.SetItem`. I don't know much about this, but this seems more promising. Perhaps we could generate an error based on the source of the setitem and type of IR. Do you know if user code should always generate the `ir.SetItem` (assuming someone doesn't try explicitly call the setitem attribute, which seems like an uncommon/unlikely pattern)?

Regarding the error, I also don't think it be reliably raised before typing. Its fine for this optimization to fail so long as you don't need the dictionary to be a literal. I don't believe you should be able to determine that before typing.

Reply:
I've been thinking about a generalisation of this problem, the question "Can we know if the the dictionary being built via updates etc is from an optimisation or from a user choice/definition"? Using this:
```python

def user_source():
    d1 = {'a': 1,
          'b': 2,
          'c': 3,
          'd': 4,
          'e': 5,
          'f': 6,
          'g': 7,
          'h': 8,
          'i': 9,
          'j': 10,
          'k': 11,
          'l': 12,
          'm': 13,
          'n': 14,
          'o': 15,
          'p': 16,
          'q': 17,}

    d2 = {"r": 18,
          "s": 19,}

    d1.update(d2)


def internal_opt_source():
    d3 = {
        "a": 1,
        "b": 2,
        "c": 3,
        "d": 4,
        "e": 5,
        "f": 6,
        "g": 7,
        "h": 8,
        "i": 9,
        "j": 10,
        "k": 11,
        "l": 12,
        "m": 13,
        "n": 14,
        "o": 15,
        "p": 16,
        "q": 17,
        "r": 18,
        "s": 19,
    }
```
it seems like the `DICT_UPDATE` only occurs in `internal_opt_source`, in the `user_source` it's evidently a method call. Combining this information with the information here: https://github.com/numba/numba/issues/7894#issuecomment-1062915850 about the source of this change from the bytecode compiler, it seems to indicate that the presence of a `DICT_UPDATE` op code can only arise in the case of a CPython induced byte code optimisation i.e. it's not reachable from user code. Is this the piece of information that helps with answering the question?

Were this information about the presence of a `DICT_UPDATE` opcode encoded into the `ir.Expr.getattr` as e.g. `getattr(dict_inst, 'update_internal')` or something similar, would that make it possible to correctly implement a standard `dict.update` as normal, then have a special overload for `dict.update_internal` that wires into `dict.update` for a standard dict but the implementation for the `LiteralStrKeyDict.update_internal` wires into an overload that raises and provides an explanation of how it should not have been reached and there must be something that's preventing the transform?

This is quite long as might be best discussed at the next Numba public meeting.

Reply:
@stuartarchibald I think this sounds like the right approach and it would also fix some of the current issues in the code (namely failing to delete the other dictionary/dictionaries). I think it would be great if we could discuss that at the next Numba meeting. I'm going to pause on any other updates until we have that discussion.